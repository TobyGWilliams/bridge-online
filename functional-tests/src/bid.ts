import { Page } from "playwright";
import getBidButtonByBid from "./get-bid-button-by-bid";
import logger from "./logger";

export default (page: Page, bid: [number, string]) => {
  logger(`bid: ${bid}`);
  return page.click(getBidButtonByBid(bid));
};
