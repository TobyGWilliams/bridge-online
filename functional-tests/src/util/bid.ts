import { Page } from "playwright";
import getBidButtonByBid from "./get-bid-button-by-bid";
import { writeLog } from "./logger";

export default async (page: Page, bid: [number, string]) => {
  writeLog(`== bid started - ${bid}`);
  await page.click(getBidButtonByBid(bid));
  writeLog(`== bid successful - ${bid}`);
  return;
};
