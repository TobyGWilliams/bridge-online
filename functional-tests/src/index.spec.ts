import { chromium, Page } from "playwright";

import addPlayer from "./util/add-player";
import seatPlayer from "./util/seat-player";
import getGameId from "./util/get-game-id";
import bid from "./util/bid";
import getGameState from "./util/get-game-state";
import getCards from "./util/get-cards";
import getState from "./util/get-state";

import {
  BUTTON_BEGIN_GAME,
  CREATE_GAME,
  INPUT_GAME_SEED,
  PLAYER1CARDS,
  PLAYER2CARDS,
  PLAYER3CARDS,
  PLAYER4CARDS,
  URL,
  BUTTON_PASS,
  GAME_STATE_LEADING_FIRST_CARD,
} from "./util/constants";

const wait = (duration: number) =>
  new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, duration);
  });

const SEED = "this is the game seed";

test("a game", async () => {
  const browser = await chromium.launch();

  try {
    const page1 = await browser.newPage();

    await page1.goto(URL);

    await page1.waitForSelector(CREATE_GAME);

    await page1.focus(INPUT_GAME_SEED);
    await page1.keyboard.type(SEED);

    await page1.click(CREATE_GAME);

    const gameId = await getGameId(page1);

    const page2 = await addPlayer(browser, gameId as string);
    const page3 = await addPlayer(browser, gameId as string);
    const page4 = await addPlayer(browser, gameId as string);

    await seatPlayer(page1, "player1Name", "north");
    await seatPlayer(page2, "player2Name", "east");
    await seatPlayer(page3, "player3Name", "south");
    await seatPlayer(page4, "player4Name", "west");

    await page1.click(BUTTON_BEGIN_GAME);

    expect(await getCards(page1)).toEqual(PLAYER1CARDS);
    expect(await getCards(page2)).toEqual(PLAYER2CARDS);
    expect(await getCards(page3)).toEqual(PLAYER3CARDS);
    expect(await getCards(page4)).toEqual(PLAYER4CARDS);

    await bid(page1, [1, "NO_TRUMPS"]);
    await bid(page2, [2, "NO_TRUMPS"]);
    await bid(page3, [3, "NO_TRUMPS"]);
    await bid(page4, [4, "NO_TRUMPS"]);

    await page1.click(BUTTON_PASS);
    await page2.click(BUTTON_PASS);
    await page3.click(BUTTON_PASS);

    await wait(1000);

    expect(await getGameState(page1)).toEqual(GAME_STATE_LEADING_FIRST_CARD);
    expect(await getGameState(page2)).toEqual(GAME_STATE_LEADING_FIRST_CARD);
    expect(await getGameState(page3)).toEqual(GAME_STATE_LEADING_FIRST_CARD);

    expect(await getState(page4)).toEqual({});

    await browser.close();
  } catch (err) {
    browser.close();
    throw err;
  }
});
