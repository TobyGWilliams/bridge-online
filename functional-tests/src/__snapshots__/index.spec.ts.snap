// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a game 1`] = `
Object {
  "east": Object {
    "availableContracts": Array [
      Array [
        5,
        "CLUB",
      ],
      Array [
        5,
        "DIAMOND",
      ],
      Array [
        5,
        "HEART",
      ],
      Array [
        5,
        "SPADE",
      ],
      Array [
        5,
        "NO_TRUMPS",
      ],
      Array [
        6,
        "CLUB",
      ],
      Array [
        6,
        "DIAMOND",
      ],
      Array [
        6,
        "HEART",
      ],
      Array [
        6,
        "SPADE",
      ],
      Array [
        6,
        "NO_TRUMPS",
      ],
      Array [
        7,
        "CLUB",
      ],
      Array [
        7,
        "DIAMOND",
      ],
      Array [
        7,
        "HEART",
      ],
      Array [
        7,
        "SPADE",
      ],
      Array [
        7,
        "NO_TRUMPS",
      ],
    ],
    "bid": "PASS",
    "currentUserAction": false,
    "isUser": false,
    "name": "player2Name",
    "wonTheContract": false,
  },
  "north": Object {
    "availableContracts": Array [
      Array [
        5,
        "CLUB",
      ],
      Array [
        5,
        "DIAMOND",
      ],
      Array [
        5,
        "HEART",
      ],
      Array [
        5,
        "SPADE",
      ],
      Array [
        5,
        "NO_TRUMPS",
      ],
      Array [
        6,
        "CLUB",
      ],
      Array [
        6,
        "DIAMOND",
      ],
      Array [
        6,
        "HEART",
      ],
      Array [
        6,
        "SPADE",
      ],
      Array [
        6,
        "NO_TRUMPS",
      ],
      Array [
        7,
        "CLUB",
      ],
      Array [
        7,
        "DIAMOND",
      ],
      Array [
        7,
        "HEART",
      ],
      Array [
        7,
        "SPADE",
      ],
      Array [
        7,
        "NO_TRUMPS",
      ],
    ],
    "bid": "PASS",
    "currentUserAction": false,
    "isUser": false,
    "name": "player1Name",
    "wonTheContract": false,
  },
  "south": Object {
    "availableContracts": Array [
      Array [
        5,
        "CLUB",
      ],
      Array [
        5,
        "DIAMOND",
      ],
      Array [
        5,
        "HEART",
      ],
      Array [
        5,
        "SPADE",
      ],
      Array [
        5,
        "NO_TRUMPS",
      ],
      Array [
        6,
        "CLUB",
      ],
      Array [
        6,
        "DIAMOND",
      ],
      Array [
        6,
        "HEART",
      ],
      Array [
        6,
        "SPADE",
      ],
      Array [
        6,
        "NO_TRUMPS",
      ],
      Array [
        7,
        "CLUB",
      ],
      Array [
        7,
        "DIAMOND",
      ],
      Array [
        7,
        "HEART",
      ],
      Array [
        7,
        "SPADE",
      ],
      Array [
        7,
        "NO_TRUMPS",
      ],
    ],
    "bid": "PASS",
    "currentUserAction": false,
    "isUser": false,
    "name": "player3Name",
    "wonTheContract": false,
  },
  "west": Object {
    "availableContracts": Array [
      Array [
        5,
        "CLUB",
      ],
      Array [
        5,
        "DIAMOND",
      ],
      Array [
        5,
        "HEART",
      ],
      Array [
        5,
        "SPADE",
      ],
      Array [
        5,
        "NO_TRUMPS",
      ],
      Array [
        6,
        "CLUB",
      ],
      Array [
        6,
        "DIAMOND",
      ],
      Array [
        6,
        "HEART",
      ],
      Array [
        6,
        "SPADE",
      ],
      Array [
        6,
        "NO_TRUMPS",
      ],
      Array [
        7,
        "CLUB",
      ],
      Array [
        7,
        "DIAMOND",
      ],
      Array [
        7,
        "HEART",
      ],
      Array [
        7,
        "SPADE",
      ],
      Array [
        7,
        "NO_TRUMPS",
      ],
    ],
    "bid": Array [
      4,
      "NO_TRUMPS",
    ],
    "cards": Array [
      Array [
        "JACK",
        "CLUB",
      ],
      Array [
        "2",
        "SPADE",
      ],
      Array [
        "ACE",
        "SPADE",
      ],
      Array [
        "2",
        "DIAMOND",
      ],
      Array [
        "4",
        "CLUB",
      ],
      Array [
        "9",
        "HEART",
      ],
      Array [
        "QUEEN",
        "SPADE",
      ],
      Array [
        "10",
        "CLUB",
      ],
      Array [
        "KING",
        "HEART",
      ],
      Array [
        "2",
        "CLUB",
      ],
      Array [
        "ACE",
        "CLUB",
      ],
      Array [
        "6",
        "SPADE",
      ],
      Array [
        "QUEEN",
        "DIAMOND",
      ],
    ],
    "currentUserAction": true,
    "isUser": true,
    "name": "player4Name",
    "wonTheContract": true,
  },
}
`;
